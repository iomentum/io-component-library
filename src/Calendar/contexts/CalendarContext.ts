import { createContext, Dispatch, SetStateAction } from "react";
import { Display, MomentRangeExtended } from "../utils";

export interface CalendarContextInterface {
  display: Display;
  setDisplay: Dispatch<SetStateAction<Display>>;
  setDate: Dispatch<SetStateAction<MomentRangeExtended>>;
}

export const CalendarContext = createContext<CalendarContextInterface>(null);
