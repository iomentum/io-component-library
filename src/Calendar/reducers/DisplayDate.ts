export enum DisplayDateType {
  UpdateStartDate,
  UpdateEndDate,
  Reset,
}

type DisplayDateAction =
  | {
      type: DisplayDateType.UpdateStartDate;
      startDate: string;
    }
  | {
      type: DisplayDateType.UpdateEndDate;
      endDate: string;
    }
  | {
      type: DisplayDateType.Reset;
      startDate: string;
      endDate: string;
    };

export interface DisplayDate {
  startDate: string;
  endDate: string;
}

const displayDateInitialState = (
  startDate: string,
  endDate: string
): DisplayDate => ({
  startDate,
  endDate,
});

export const displayDateReducer = (
  state: DisplayDate,
  action: DisplayDateAction
): DisplayDate => {
  switch (action.type) {
    case DisplayDateType.UpdateStartDate:
      return { ...state, startDate: action.startDate };
    case DisplayDateType.UpdateEndDate:
      return { ...state, endDate: action.endDate };
    case DisplayDateType.Reset:
      return displayDateInitialState(action.startDate, action.endDate);
  }
};
