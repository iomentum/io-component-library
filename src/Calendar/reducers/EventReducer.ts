import { getFormattedDate } from '../utils/dateUtils';

export enum EventType {
  UpdateTitle,
  UpdateStartDate,
  UpdateEndDate,
  UpdateStartHour,
  UpdateEndHour,
  UpdateEvent,
  Reset,
}

export type EventAction =
  | {
      type: EventType.UpdateTitle;
      title: string;
    }
  | {
      type: EventType.UpdateStartDate;
      startDate: Date;
    }
  | {
      type: EventType.UpdateEndDate;
      endDate: Date;
    }
  | {
      type: EventType.UpdateStartHour;
      startHour: string;
    }
  | {
      type: EventType.UpdateEndHour;
      endHour: string;
    }
  | {
      type: EventType.UpdateEvent;
      uuid: string;
      title: string;
      startDate: Date;
      endDate: Date;
      startHour: string;
      endHour: string;
      metadata: Record<string, unknown>;
    };

export interface Event {
  uuid: string;
  title: string;
  startDate: Date;
  displayStartDate: string;
  endDate: Date;
  displayEndDate: string;
  startHour: string;
  endHour: string;
  metadata: Record<string, unknown>;
}

export const eventReducer = (state: Event, action: EventAction): Event => {
  switch (action.type) {
    case EventType.UpdateTitle:
      return { ...state, title: action.title };
    case EventType.UpdateStartDate:
      return {
        ...state,
        startDate: action.startDate,
        displayStartDate: getFormattedDate(action.startDate),
      };
    case EventType.UpdateEndDate:
      return {
        ...state,
        endDate: action.endDate,
        displayEndDate: getFormattedDate(action.endDate),
      };
    case EventType.UpdateStartHour:
      return { ...state, startHour: action.startHour };
    case EventType.UpdateEndHour:
      return { ...state, endHour: action.endHour };
    case EventType.UpdateEvent:
      return {
        uuid: action.uuid,
        title: action.title,
        startDate: action.startDate,
        displayStartDate: getFormattedDate(action.startDate),
        endDate: action.endDate,
        displayEndDate: getFormattedDate(action.endDate),
        startHour: action.startHour,
        endHour: action.endHour,
        metadata: action.metadata,
      };
    default:
      throw Error('Unknow EventType');
  }
};
