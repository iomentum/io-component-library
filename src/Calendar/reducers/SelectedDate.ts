export enum SelectedDateType {
  UpdateStartDate,
  UpdateEndDate,
  Reset,
}

type DisplayDateAction =
  | {
      type: SelectedDateType.UpdateStartDate;
      startDate: Date;
    }
  | {
      type: SelectedDateType.UpdateEndDate;
      endDate: Date;
    }
  | {
      type: SelectedDateType.Reset;
      startDate: Date;
      endDate: Date;
    };

interface SelectedDate {
  startDate: Date;
  endDate: Date;
}

const selectedDateInitialState = (
  startDate: Date,
  endDate: Date
): SelectedDate => ({
  startDate,
  endDate,
});

export const selectedDateReducer = (
  state: SelectedDate,
  action: DisplayDateAction
): SelectedDate => {
  switch (action.type) {
    case SelectedDateType.UpdateStartDate:
      return { ...state, startDate: action.startDate };
    case SelectedDateType.UpdateEndDate:
      return { ...state, endDate: action.endDate };
    case SelectedDateType.Reset:
      return selectedDateInitialState(action.startDate, action.endDate);
  }
};