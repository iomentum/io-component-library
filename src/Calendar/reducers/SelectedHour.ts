export enum SelectedHourType {
  UpdateStartHour,
  UpdateEndHour,
  Reset,
}

type DisplayHourAction =
  | {
      type: SelectedHourType.UpdateStartHour;
      startHour: string;
    }
  | {
      type: SelectedHourType.UpdateEndHour;
      endHour: string;
    }
  | {
      type: SelectedHourType.Reset;
      startHour: string;
      endHour: string;
    };

interface SelectedHour {
  startHour: string;
  endHour: string;
}

const selectedHourInitialState = (
  startHour: string,
  endHour: string
): SelectedHour => ({
  startHour,
  endHour,
});

export const selectedHourReducer = (
  state: SelectedHour,
  action: DisplayHourAction
): SelectedHour => {
  switch (action.type) {
    case SelectedHourType.UpdateStartHour:
      return { ...state, startHour: action.startHour };
    case SelectedHourType.UpdateEndHour:
      return { ...state, endHour: action.endHour };
    case SelectedHourType.Reset:
      return selectedHourInitialState(action.startHour, action.endHour);
  }
};
