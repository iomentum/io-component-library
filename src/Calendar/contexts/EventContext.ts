import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, Event } from '../reducers/EventReducer';

export interface EventContextInterface {
  eventsCollection: Event[];
  setEventsCollection: Dispatch<SetStateAction<Event[]>>;
  eventManagementOpened: boolean;
  setEventManagementOpened: Dispatch<SetStateAction<boolean>>;
  event: Event;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventContext = createContext<EventContextInterface>(null);
