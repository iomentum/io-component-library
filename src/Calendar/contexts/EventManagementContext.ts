import { createContext, Dispatch, SetStateAction } from 'react';
import { EventAction, EventModel } from '../reducers/EventReducer';

export interface EventManagementContext {
  fullDayEvent: boolean;
  event: EventModel;
  setFullDayEvent: Dispatch<SetStateAction<boolean>>;
  dispatchEvent: Dispatch<EventAction>;
}

export const EventManagementContext = createContext<EventManagementContext>(null);
