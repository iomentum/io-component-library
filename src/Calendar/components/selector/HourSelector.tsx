import React, { useContext } from 'react';
import { TextField } from '@material-ui/core';
import { EventContext } from '../../contexts/EventContext';
import { EventType } from '../../reducers/EventReducer';
import { Row, Gap } from '../common.style';

export enum EventHourType {
  Start,
  End,
}

export const computeNewHours = (
  startHour: string,
  endHour: string,
  hourChanged: EventHourType
): [string, string] => {
  if (startHour > endHour) {
    if (startHour > '23:00') {
      return [startHour, '23:59'];
    }
    if (endHour < '01:00') {
      return ['00:00', endHour];
    }
    if (hourChanged === EventHourType.Start) {
      const [tensHours, unitHours, separator, tensMinutes, unitMinutes] = startHour.split('');
      return [startHour, `${tensHours}${+unitHours + 1}${separator}${tensMinutes}${unitMinutes}`];
    }
    if (hourChanged === EventHourType.End) {
      const [tensHours, unitHours, separator, tensMinutes, unitMinutes] = endHour.split('');
      return [`${tensHours}${+unitHours - 1}${separator}${tensMinutes}${unitMinutes}`, endHour];
    }
  }
  return [startHour, endHour];
};

export const HourSelector = () => {
  const { event, dispatchEvent } = useContext(EventContext);

  return (
    <Row>
      <TextField
        fullWidth
        label="Start"
        type="time"
        value={event.startHour}
        inputProps={{
          'data-testid': 'startTime',
        }}
        onChange={(onChangeEvent) => {
          const [startHour, endHour] = computeNewHours(
            onChangeEvent.target.value,
            event.endHour,
            EventHourType.Start
          );
          dispatchEvent({
            type: EventType.UpdateHours,
            startHour,
            endHour,
          });
        }}
      />
      <Gap />
      <TextField
        fullWidth
        label="End"
        type="time"
        value={event.endHour}
        inputProps={{
          'data-testid': 'endTime',
        }}
        onChange={(onChangeEvent) => {
          const [startHour, endHour] = computeNewHours(
            event.startHour,
            onChangeEvent.target.value,
            EventHourType.End
          );
          dispatchEvent({
            type: EventType.UpdateHours,
            startHour,
            endHour,
          });
        }}
      />
    </Row>
  );
};
