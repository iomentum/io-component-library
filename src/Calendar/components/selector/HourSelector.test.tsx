/**
 * @jest-environment jsdom
 */

import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { EventManagementContext } from '../../contexts/EventManagementContext';
import { HourSelector, computeStartHourChanges, computeEndHourChanges } from './HourSelector';
import { EventType } from '../../reducers/EventReducer';

const eventManagementContextMock = (component, { providerValue, ...renderOptions }) =>
  render(
    <EventManagementContext.Provider value={providerValue}>
      {component}
    </EventManagementContext.Provider>,
    renderOptions
  );

describe('HourSelector component', () => {
  describe('@snapshots', () => {
    it('should match with the previous HourSelector', () => {
      const providerValue = {
        event: {
          startHour: '00:00',
          endHour: '01:00',
        },
        dispatchEvent: jest.fn(),
      };
      const { asFragment } = eventManagementContextMock(<HourSelector />, { providerValue });

      expect(asFragment()).toMatchSnapshot('HourSelector snapshot');
    });
  });

  describe('@events', () => {
    it('should be dispatched with the right startHour', () => {
      const providerValue = {
        event: {
          startHour: '00:00',
          endHour: '01:00',
        },
        dispatchEvent: jest.fn(),
      };
      const component = eventManagementContextMock(<HourSelector />, {
        providerValue,
      });

      const input = component.getByTestId('startTime') as HTMLInputElement;
      fireEvent.change(input, { target: { value: '01:10' } });

      expect(providerValue.dispatchEvent).toHaveBeenCalledWith({
        type: EventType.UpdateStartHour,
        startHour: '01:10',
      });
    });

    it('should be dispatched with the right endHour', () => {
      const providerValue = {
        event: {
          startHour: '00:00',
          endHour: '01:00',
        },
        dispatchEvent: jest.fn(),
      };
      const component = eventManagementContextMock(<HourSelector />, {
        providerValue,
      });

      const input = component.getByTestId('endTime') as HTMLInputElement;
      fireEvent.change(input, { target: { value: '01:10' } });

      expect(providerValue.dispatchEvent).toHaveBeenCalledWith({
        type: EventType.UpdateEndHour,
        endHour: '01:10',
      });
    });
  });
});

describe('computeHourChanges functions', () => {
  describe('computeStartHourChanges function', () => {
    it('should not trigger dispatchEvent, 1st case', () => {
      const dispatchEvent = jest.fn();
      computeStartHourChanges('00:00', '00:50', dispatchEvent);

      expect(dispatchEvent).not.toHaveBeenCalled();
    });

    it('should not trigger dispatchEvent, 2nd case', () => {
      const dispatchEvent = jest.fn();
      computeStartHourChanges('23:15', '23:16', dispatchEvent);

      expect(dispatchEvent).not.toHaveBeenCalled();
    });

    it('should update endHour with +1 hour', () => {
      const dispatchEvent = jest.fn();
      computeStartHourChanges('01:00', '00:50', dispatchEvent);

      expect(dispatchEvent).toHaveBeenCalledWith({
        type: EventType.UpdateEndHour,
        endHour: '02:00',
      });
    });

    it('should update endHour with 23:59', () => {
      const dispatchEvent = jest.fn();
      computeStartHourChanges('23:10', '00:50', dispatchEvent);

      expect(dispatchEvent).toHaveBeenCalledWith({
        type: EventType.UpdateEndHour,
        endHour: '23:59',
      });
    });
  });

  describe('computeEndHourChanges function', () => {
    it('should not trigger dispatchEvent, 1st case', () => {
      const dispatchEvent = jest.fn();
      computeEndHourChanges('00:00', '00:50', dispatchEvent);

      expect(dispatchEvent).not.toHaveBeenCalled();
    });

    it('should not trigger dispatchEvent, 2nd case', () => {
      const dispatchEvent = jest.fn();
      computeEndHourChanges('23:15', '23:16', dispatchEvent);

      expect(dispatchEvent).not.toHaveBeenCalled();
    });

    it('should update startHour with +1 hour', () => {
      const dispatchEvent = jest.fn();
      computeEndHourChanges('02:50', '02:00', dispatchEvent);

      expect(dispatchEvent).toHaveBeenCalledWith({
        type: EventType.UpdateStartHour,
        startHour: '01:00',
      });
    });

    it('should update startHour with 00:00', () => {
      const dispatchEvent = jest.fn();
      computeEndHourChanges('23:10', '00:50', dispatchEvent);

      expect(dispatchEvent).toHaveBeenCalledWith({
        type: EventType.UpdateStartHour,
        startHour: '00:00',
      });
    });
  });
});
