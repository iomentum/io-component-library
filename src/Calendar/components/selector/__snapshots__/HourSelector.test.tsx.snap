// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HourSelector component @snapshots should match with the previous HourSelector: HourSelector snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
    >
      Start
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        data-testid="startTime"
        type="time"
        value="00:00"
      />
    </div>
  </div>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
      data-shrink="true"
    >
      End
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        data-testid="endTime"
        type="time"
        value="01:00"
      />
    </div>
  </div>
</DocumentFragment>
`;
